File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # if not head:
        #     return False
        # if "visited" in head and head.visited:
        #     return True
        # head.visited = True
        # return self.hasCycle(head.next)
        visited = set()

        while head:
            if head in visited:
                return True
            visited.add(head)
            head = head.next
            
        return False
        

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null) return null;
        return reverse(head, null);
    }
    private ListNode reverse(ListNode head, ListNode prev) {
        ListNode newHead = head;
        if (newHead.next != null) {
            newHead = reverse(head.next, head);
        }
        head.next = prev;
        return newHead;
    }
}
